### PR TITLE
New strategy to share implementation between sync and Lwt

### PR DESCRIPTION
**Pull request by mjambon**
Most modules are now shared via symlinks except for the `Monad` module. This makes the source code straightforward and still avoids the complexity of multifile functors. As a side effect, the `testo` and `testo-lwt` libraries are now independent of one another.

This is a preliminary to #57 which would otherwise require code duplication.

I need to check that this works in semgrep. I removed `Testo.update_func` in the process, which was a function needed to convert Lwt tests to synchronous tests. I'm not sure if it's really needed or what to do about it. Update: the solution I found is to not use `Testo.update_func`: https://github.com/semgrep/semgrep/pull/10147

PR checklist:

- [x] Purpose of the code is evident to future readers
- [x] Tests are included or a PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [ ] A changelog entry was added to `CHANGES.md` for any user-facing change

Check out [`CONTRIBUTING.md`](https://github.com/semgrep/testo/blob/main/CONTRIBUTING.md) for more details.

### CODE REVIEW

File: core/Run.ml [hunk 1]
```diff
@@ -468,7 +470,7 @@ let ends_with_newline str =
   str <> "" && str.[String.length str - 1] = '\n'
 
 let print_status ~highlight_test ~always_show_unchecked_output
-    (((test : _ T.test), (status : T.status), sum) as test_with_status)=
+    (((test : T.test), (status : T.status), sum) as test_with_status)=
```
Comment (amchiclet) [inline core/Run.ml:473]: Does ocamlformat allow no space before `=`?

**Review by amchiclet — APPROVED** [review]
I don't have much clarity about how the symlinks are set up, but maybe I just need to spend time to understand the overall architecture of Testo.

I only commented on very minor things. You can wait for Pad in case he has more meaningful feedback than I do.

File: core/Run.ml (hunk 1 above)
Comment (mjambon) [reply core/Run.ml:473]: ocamlformat doesn't run on this repo. It's weird that I didn't see the missing space, I'm usually annoyed by these things.

**Comment by mjambon** [comment]
@amchiclet 
> I don't have much clarity about how the symlinks are set up

I'll add some kind of readme alongside the symlinks since this is an unusual setup.

File: dune-project [hunk 2]
```diff
@@ -34,13 +30,15 @@ new subcommands for capturing, checking, and approving the output of tests.")
...
 (package
   (name testo-lwt)
-  (synopsis "Extension of the Alcotest library for Lwt")
+  (synopsis "Test framework for OCaml, Lwt variant")
   (description "\
 Use this if the tests return Lwt promises and you can't make them synchronous
 by wrapping them in 'Lwt_main.run'.")
   (depends
     (ocaml (>= 4.08.0))
+    alcotest
```
Comment (aryx) [inline dune-project:39]: maybe explain here why we still depend on alcotest?

File: testo-lwt.opam [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # This file is generated by dune, edit dune-project instead
 opam-version: "2.0"
-synopsis: "Extension of the Alcotest library for Lwt"
+synopsis: "Test framework for OCaml, Lwt variant"
 description: """
 Use this if the tests return Lwt promises and you can't make them synchronous
```
Comment (aryx) [inline testo-lwt.opam:5]: In which case you can't wrap them in Lwt_main.run ?

**Review by aryx — COMMENTED** [review]
LGTM.
Glad we're getting a simpler Testo.mli without those polymorphic types.

File: testo-lwt.opam (hunk 3 above)
Comment (mjambon) [reply testo-lwt.opam:5]: right

File: core/Promise.mli [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+val return : 'a -> 'a t
+val bind : 'a t -> ('a -> 'b t) -> 'b t
+val catch :
+  (unit -> 'a t) ->
+  (exn -> Printexc.raw_backtrace -> 'a t) ->
+  'a t
+
+module Operators : sig
+  (* The bind operator *)
+  val ( >>= ) : 'a t -> ('a -> 'b t) -> 'b t
```
[the remark this answers is not in the document]
Comment (mjambon) [reply core/Promise.mli:16]: I'll look into some day.

File: dune-project (hunk 2 above)
Comment (mjambon) [reply dune-project:39]: ok

**Pull request merged**